asf_public_discourse_home_decarbonisation/pipeline/faqs_identification/extract_questions.py

```diff
@@ -0,0 +1,445 @@
+            question_pattern, sentence, flags=re.IGNORECASE
+        )
+        all_questions.extend(potential_questions)
+    return all_questions
+
+
+def filter_out_short_questions(
+    questions: List[str], num_of_words: int = 5, min_length: int = 5
+) -> List[str]:
+    """
+    Filters out short questions from a list of questions based on the number of words in the question as well as the minimum length of the question.
+
+    Args:
+        questions (List[str]): The list of questions to filter.
+        num_of_words (int): The minimum number of words a question should have to be included.
+        min_length (int): The minimum length a question should have to be included.
+
+    Returns:
+        List[str]: The filtered list of questions.
+    """
+    filtered_questions = []
+    # Skip questions that are 'num_of_words' words or less and contain a question word
+    for question in questions:
+        words = nltk.word_tokenize(question)
+        if len(words) <= num_of_words and any(
+            word.lower() in question_words for word in words
+        ):
+            continue
+        # Add sentence to all_questions if it's at least min_length long and not just question marks with optional spaces.
+        if len(question) >= min_length and not re.fullmatch(r"\?+\s*", question):
+            filtered_questions.append(question)
+    return filtered_questions
+
+
+def extract_idk(sentences: List[str], inclusion_phrases: List[str]) -> List[str]:
+    """
+    Extracts and returns a list of sentences with specific "I don't know" phrases.
+
+    The function searches for sentences that contain any of a predefined set of phrases indicating uncertainty or lack of knowledge (e.g., "don’t know", "do not know", "do not know how to"). It then puts these into a list.
+
+    This can be particularly useful for analysing text where it's important to both identify expressions of uncertainty and examine the context without these expressions.
+
+    Args:
+        sentences (List[str]): A list of sentences to be evaluated. Each item in the list should be a string representing a single sentence.
+        inclusion_phrases (List[str]): A list of inclusion phrases. Each item in the list should be a string representing a phrase to search for in the sentences.
+
+    Returns:
+        idk_phrases (List[str]):
+            - This list contains sentences that include any of the predefined "don't know" phrases.
+
+    Note:
+        - The function is case-insensitive when searching for the predefined phrases.
+    """
+    idk_phrases = []
+    for sentence in sentences:
+        if any(phrase in sentence.lower() for phrase in inclusion_phrases):
+            idk_phrases.append(sentence)
+    return idk_phrases
+
+
+def extract_X_from_idk(
+    idk_phrases: List[str], inclusion_phrases: List[str]
+) -> List[str]:
+    """
+    Extracts and returns a list of sentences without the inclusion phrases found in the "I don't know" phrases.
```

Comment:
Just thought we could make this a bit more explicit.

```suggestion
    Extracts and returns a list of expressions coming after the inclusion_phrases in sentences.
    Example:
         - idk_phrases: ["honestly I don't know what a heat pump is", "i have asked multiple people, but still don't know how what to do to install a heat pump in my home"
         - inclusion_phrases: ["don't know how to", "don't know"]
         - Returns: ["what a heat pump is", "what to do to install a heat pump in my home"]
```